File: core/src/main/java/com/xmartlabs/bigbang/core/module/SessionInterceptor.java
package com.xmartlabs.bigbang.core.module;

import android.support.annotation.NonNull;

import com.xmartlabs.bigbang.core.Injector;
import com.xmartlabs.bigbang.core.providers.AccessTokenProvider;

import java.io.IOException;

import javax.inject.Inject;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public class SessionInterceptor implements Interceptor {
  @Inject
  AccessTokenProvider accessTokenProvider;

  public SessionInterceptor() {
    Injector.inject(this);
  }

  @NonNull
  @Override
  public Response intercept(@NonNull Chain chain) throws IOException {
    Request request = accessTokenProvider.provideEntity()
        .map(accessToken -> chain.request()
            .newBuilder()
            .addHeader(accessTokenProvider.provideAccessTokenHeaderKey(), accessToken)
            .build()
        ).orElse(chain.request());
    return chain.proceed(request);
  }
}

File: core/src/main/java/com/xmartlabs/bigbang/core/module/SessionInterceptor.kt
package com.xmartlabs.bigbang.core.module

import com.xmartlabs.bigbang.core.providers.AccessTokenProvider
import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException
import javax.inject.Inject

class SessionInterceptor @Inject constructor(private val accessTokenProvider: AccessTokenProvider) : Interceptor {

  @Throws(IOException::class)
  override fun intercept(chain: Interceptor.Chain): Response {
    val request = accessTokenProvider.provideEntity()
        ?.let {
          chain.request()
              .newBuilder()
              .addHeader(accessTokenProvider.provideAccessTokenHeaderKey(), it)
              .build()
        }
    return chain.proceed(request ?: chain.request())
  }
}
